PROJ-147 Camera handling (#83)
SW805F20/report: sections/sprint-4/camera-handling.tex

\section{Camera handling}

> This section will explain how the camera in the game is handled.

Reviewer: ```suggestion
There were some complications with handling the camera in Unity.
```

Reviewer: Ved ikke helt om det lige præcis skal være denne formulering, som skal bruges som introduktion til afsnittet, men det er nok ikke nødvendigt at forklare, at camera handling vil forklares i denne sektion

> Since the playing field can have a varying size the camera needs to be able to fit the whole playing field on the screen so that it can be rendered.

Reviewer 2: ```suggestion
Since the playing field can have a varying size the camera needs to be able to fit the whole playing field on the screen such that it can be rendered.
```

> Another problem is the usage of VR headsets that allow the player to turn the camera's view by looking in different directions.

Reviewer 2: ```suggestion
Another problem is the usage of VR headsets that allow the player to manipulate the camera's view by looking in different directions.
```

> This causes the playing field to go out of the player's view which is a problem since they can then no longer see their position.

Reviewer 2: ```suggestion
This can cause the playing field to go out of the player's view, which is a problem since they can then no longer see their position.
```

\\\\

> The way this is solved is by always having the playing field and its attached game objects face in the direction of the camera.

Reviewer 2: ```suggestion
The way this is solved is by always having the playing field and its attached game objects face the direction of the camera.
```

This is achieved by giving the playing field the same rotation as the camera.
By doing this, the player will always see the playing field in the same position in the 3D space no matter which direction they look.

> But this solution is not enough since the playing field will not have the same position in the 3D space as the camera.

Reviewer 2: ```suggestion
But this solution is not enough, since the playing field will not have the same position in the 3D space as the camera.
```

It must be a certain distance away from the camera, based on its width and height, to allow it to be rendered. 

> Because of this, a new position must be calculated for the playing field in relation to the cameras direction. 

Reviewer 2: ```suggestion
Because of this, a new position must be calculated for the playing field in relation to the direction of the camera. 
```

\begin{lstlisting}[caption={Calculating and setting the position and rotation of the playing field}, captionpos=b,language=C,label={lst:camera_position}]
    newPosition = playingFieldTransform.transform.position + playingFieldTransform.transform.forward * distanceFromCamera;
    transform.position = newPosition;
    transform.rotation = mainCamera.transform.localRotation; 
\end{lstlisting}

> The new position is calculated by adding the playing field's initial position relative to the camera \texttt{playingFieldTransform.transform.position} with an offset vector.

Reviewer 2: det er ligesom om der mangler et eller andet her mellem camera og så den der texttt

> The new position is calculated by adding the playing field's initial position relative to the camera \texttt{playingFieldTransform.transform.position} with an offset vector.

Reviewer: ```suggestion
As seen on \autoref{lst:camera_position} the new position is calculated by adding the playing field's initial position relative to the camera \texttt{playingFieldTransform.transform.position} with an offset vector.
```

> The offset vector is calculated by multiplying \texttt{playingFieldTransform.transform.forward}, which is a normalized vector pointing in the camera's looking direction, with a certain distance \texttt{distanceFromCamera} that we want the field to be from the camera.

Reviewer 2: ```suggestion
The offset vector is calculated by multiplying \texttt{playingFieldTransform.transform.forward}, which is a normalized vector pointing in the camera's looking direction, with a certain distance defined through the \texttt{distanceFromCamera} variable, which is how far the field should be from the camera.
```

These calculations are done in the \texttt{Update} method which is called for every frame rendered in the game.
\\\\

> For the \texttt{distanceFromCamera} calculation, the camera needs to be far enough away from the field to fit the it in its far clipping view.

Reviewer 2: ```suggestion
For the \texttt{distanceFromCamera} calculation, the camera needs to be far enough away from the field to fit it in its far clipping view.
```

Reviewer 2: ved ikke om der mnagler lidt til termet clipping view

\begin{lstlisting}[caption={Calculation of the distance between the camera and field}, captionpos=b,language=C,label={lst:distanceFromCamera}]
    distanceFromCamera = 1.2f * size * 0.5f / Mathf.Tan(mainCamera.fieldOfView * 0.5f * Mathf.Deg2Rad);
\end{lstlisting}

> This is a simple trigonomotry problem where a triangle is formed between the camera, the center of the field and the top of the field.

Reviewer: ```suggestion
This is a simple trigonometry problem where a triangle is formed between the camera, the center of the field and the top of the field.
```

The height of the field and the angle of the camera's field of view are the known values, and the distance between the camera and the center of the field can be found. 

> Half of the side opposite to the camera is divided with the tangent function of half of the field of view angle converted radian.

Reviewer 2: ```suggestion
Half of the side opposite to the camera is divided with the tangent function of half of the field of view angle converted to radians.
```
